hips/hip-9997.md

```diff
@@ -0,0 +1,197 @@
+
+In addition, there has not yet been a clear response to the following questions concerning Helm's OCI support:
+
+- What is the relationship between a chart version and registry tag?
+- Will this work with `helm install` / `helm upgrade` / `helm dependency`?
+- How will provenance files (`.prov`) be supported?
+- When will the experimental flag be removed?
+
+## Rationale
+
+The true value of leveraging OCI specs has very little to do with the command-line experience (i.e. Docker). OCI registries provide a common API for all types of packages, and address several security and scalability concerns. Additionally, many companies and organizations have invested in the infrastructure surrounding their container registry. Storing Helm charts in a registry reduces the number of moving parts.
+
+By making Helm's OCI support more closely aligned with the way that Helm currently works, users will have a more stable experience and still benefit from the advantages of OCI.
+
+## Specification
+
+The specification for this HIP is broken into six (6) major sections:
+
+1. Implement `Getter` and introduce `Pusher`
+2. Support for provenance files
+3. Chart versions == OCI reference tags
+4. Chart names == OCI reference basenames
+5. Cache is removed
+6. `helm chart` is removed, integrated into rest of CLI
+7. Experimental until clear messaging from OCI
+
+### 1. Implement `Getter` and introduce `Pusher`
+
+As of yet, the code related to registry support has been written standalone (see [internal/experimental/registry](https://github.com/helm/helm/tree/master/internal/experimental/registry)).
+
+The act of downloading a chart from an OCI registry should mimic what is already possible using downloader plugins. Using the protocol prefix `oci://` in any place where chart repos are referenced should work just by implementing a `Getter` for OCI.
+
+Currently there is no Helm-specific way to upload chart packages. Using the same model of `Getter`, a new interface called `Pusher` should be introduced, with a single built-in implementation (OCI). This opens the door for plugins to implement their own upload mechanism, and for Helm to add a new top-level upload command `helm push`.
+
+### 2. Support for provenance files
+
+Helm currently has the ability to verify package signatures, assuming the presence of a file suffixed with `.prov` sitting next to a chart `.tgz` in a repository.
+
+Support for this method of signature validation should be carried over into OCI storage. As the format of the OCI manifest is custom to Helm, Helm can also choose to modify the resulting manifest on upload when a chart is being signed (e.g. `helm push --sign`).
+
+As far as the low-level details, the `.prov` file will simply be stored as a second layer on the manifest. If running `helm pull --verify oci://...`, the second layer (layer at index 1) will be assumed to be the provenance file. The first layer is always the chart `.tgz` itself.
```

Comment:
yes, I think that's more reasonable. However what does it mean if these layers are ordered differently than how helm pushes them?

Reply:
I think that should be fine, so long as there are only 1-2 layers, and that if there's one layer, it's the chart media type, else it's the chart and the provenance media types.

I'm not sure if we should consider 3+ layers found as an error case or not. What do you think?

Reply:
The ambiguity of this is why I thought it made sense to validate the order. We can validate the order and the content types.

There might be a use case of other tools appending other layers.. Should we just say, the first one w/ the content mediatype is treated as the tarball? Maybe loop through and make sure its the only one, otherwise error out? (same with provenance)

Reply:
> Should we just say, the first one w/ the content mediatype is treated as the tarball? Maybe loop through and make sure its the only one, otherwise error out? (same with provenance)

That seems sensible.

Reply:
added language to reflect